File: src/SimpleLevelEditor.Formats/Types/EntityConfig/EntityConfigData.fs
namespace SimpleLevelEditor.Formats.Types.EntityConfig

type EntityConfigData =
    { mutable ModelPaths: string list
      mutable TexturePaths: string list
      mutable Entities: EntityDescriptor list }

    static member CreateDefault() =
        { ModelPaths = []
          TexturePaths = []
          Entities = [] }

    member this.DeepCopy() =
        { this with
            ModelPaths = this.ModelPaths |> List.map id
            TexturePaths = this.TexturePaths |> List.map id
            Entities = this.Entities |> List.map (_.DeepCopy()) }

    member this.AddModel(model: string) =
        if not (this.ModelPaths |> List.exists (fun v -> v = model)) then
            this.ModelPaths <- this.ModelPaths @ [ model ]

    member this.AddTexture(texture: string) =
        if not (this.TexturePaths |> List.exists (fun v -> v = texture)) then
            this.TexturePaths <- this.TexturePaths @ [ texture ]

    member this.AddEntity(entity: EntityDescriptor) =
        this.Entities <- this.Entities @ [ entity ] // TODO: Check for duplicates

    member this.RemoveModel(model: string) =
        this.ModelPaths <- this.ModelPaths |> List.filter (fun v -> v <> model)

    member this.RemoveTexture(texture: string) =
        this.TexturePaths <- this.TexturePaths |> List.filter (fun v -> v <> texture)

    member this.RemoveEntity(entity: EntityDescriptor) =
        this.Entities <- this.Entities |> List.filter (fun v -> v <> entity)

File: src/SimpleLevelEditor.Formats.EntityConfig/EntityConfigData.cs
using System.Diagnostics.CodeAnalysis;

namespace SimpleLevelEditor.Formats.EntityConfig;

public sealed record EntityConfigData
{
	public required List<string> ModelPaths { get; init; }

	public required List<string> TexturePaths { get; init; }

	public required List<EntityDescriptor> Entities { get; init; }

	[SetsRequiredMembers]
	public EntityConfigData(List<string> modelPaths, List<string> texturePaths, List<EntityDescriptor> entities)
	{
		ModelPaths = modelPaths;
		TexturePaths = texturePaths;
		Entities = entities;
	}

	public static EntityConfigData CreateDefault()
	{
		return new EntityConfigData([], [], []);
	}

	public EntityConfigData DeepCopy()
	{
		return new EntityConfigData([..ModelPaths], [..TexturePaths], [..Entities.Select(e => e.DeepCopy())]);
	}

	public void AddModelPath(string path)
	{
		if (!ModelPaths.Contains(path))
			ModelPaths.Add(path);
	}

	public void AddTexturePath(string path)
	{
		if (!TexturePaths.Contains(path))
			TexturePaths.Add(path);
	}

	public void AddEntity(EntityDescriptor entity)
	{
		// TODO: Check for duplicate names.
		Entities.Add(entity);
	}

	public void RemoveModelPath(string path)
	{
		ModelPaths.Remove(path);
	}

	public void RemoveTexturePath(string path)
	{
		TexturePaths.Remove(path);
	}

	public void RemoveEntity(EntityDescriptor entity)
	{
		Entities.Remove(entity);
	}
}
